### PR TITLE
Mplex timeout & size limits

### PR DESCRIPTION
**Pull request by sinkingsugar**
Partial small addition in order to close https://github.com/status-im/nim-libp2p/issues/45

@dryajov 
What's the current best practice for error handling btw? such this max size
Also, I added a WIP test and the curious thing is that it makes another test fail:
`[1m[31m  [FAILED] [0me2e - multiple read/write streams[0m`
Trying to figure things out now :)

### CODE REVIEW

**Comment by sinkingsugar** [comment]
Relevant unrelated test failure line:
https://travis-ci.org/status-im/nim-libp2p/jobs/646738805#L1095

**Comment by dryajov** [comment]
> What's the current best practice for error handling btw? such this max size

I've been using standard exception handling practices. In this case the exception would break the mplex loop and eventually close the connection in the finally clause.

For a broader perspective, we're still working out what our prefered approach is for all status nim projects. There are so far two proposals, use exceptions or the Option/Result pattern, most of our code still uses exceptions pretty extensively however.

**Comment by sinkingsugar** [comment]
Yup, that was my intention with the exception, do you think it should take another path rather than closing? I think that since those numbers are there from the spec, this error should be indeed a early failure so that can be fixed as it's a bug.

Hmm I saw Option/Result.. honestly I use them a lot in rust, not a huge fan.. but I am a pro `NULL` guy 😃 
I think exceptions are more the nim way honestly.
Option/Result are part of the rust language, the standard library and the compiler in so many aspects that is not exactly possible in nim imho.

**Comment by dryajov** [comment]
> Yup, that was my intention with the exception, do you think it should take another path rather than closing? I think that since those numbers are there from the spec, this error should be indeed a early failure so that can be fixed as it's a bug.

I think closing is the safest thing to do even with timeouts/max sizes, we never know how much data has been already consumed from the internal buffer so the connection is effectively in an inconsistent state.

**Comment by dryajov** [comment]
> Hmm I saw Option/Result.. honestly I use them a lot in rust, not a huge fan.. but I am a pro NULL guy 😃
I think exceptions are more the nim way honestly.
Option/Result are part of the rust language, the standard library and the compiler in so many aspects that is not exactly possible in nim imho.

Yeah, honestly I think exceptions are good enough, other are pretty strongly on the other end of the spectrum. I would say that from using exceptions in Nim, I've so far haven't ran into many issues, except perhaps one, where result can be mangled/leak after an the exception is thrown (side note, we should audit the code for such cases).

**Comment by sinkingsugar** [comment]
@dryajov not sure you noticed but we got something scary going on for example (talking of leaks etc such), will look into it now, but see that Linux fails Windows does not .. (and mac does fail)

I added a new test that allocates "a lot" (1mb) of memory.. I think it's causing some memory issues to pop likely... I hope I just made a huge mistake and nothing is wrong...

**Comment by sinkingsugar** [comment]
Cancel what I said partially ! issue only happens with clang on mac...

**Comment by sinkingsugar** [comment]
Found the issue, on mac basically the next test is receiving the big buffer, correctly throwing (altho the error message should be more specific)!
I suspect that this is not expected tho, will investigate further, luckily it's on our side.

File: libp2p/muxers/mplex/mplex.nim [hunk 1]
```diff
@@ -103,8 +103,12 @@ method handle*(m: Mplex) {.async, gcsafe.} =
         of MessageType.MsgIn, MessageType.MsgOut:
           trace "pushing data to channel", id = id,
                                            initiator = initiator,
-                                           msgType = msgType
+                                           msgType = msgType,
+                                           size = data.len
 
+          if data.len > MaxMsgSize:
+             raise newException(CatchableError,
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply libp2p/muxers/mplex/mplex.nim]: Incidentally this actually existed `newLPStreamLimitError` I will just use this one exception!

**Comment by sinkingsugar** [comment]
Fixed the issue, also I noticed that all the tests were inconsistent in terms of actually waiting for the async events to be processed... so I'm proposing to use this pattern in every test:
https://github.com/sinkingsugar/nim-libp2p/commit/fb5fcc6c8b4775e9ed4466cbb2066d7f82a403b6

File: tests/testmplex.nim [hunk 2]
```diff
@@ -140,6 +144,48 @@ suite "Mplex":
...
+
+      let mplexDial = newMplex(conn)
+      let stream  = await mplexDial.newStream()
+      var bigseq = newSeqOfCap[uint8](MaxMsgSize + 1)
+      for _ in 0..<MaxMsgSize:
+        bigseq.add(uint8(rand(int('A')..int('z'))))
+      await stream.writeLp(bigseq)
+      await lock or timeout
+      check timeout.finished # this test has to timeout!
+      lock.cancel();
```
Comment (arnetheduck) [inline tests/testmplex.nim]: since the topic of exception safety was raised, this is a school-book example of broken code - on any exception (and there are plenty of them above, even if there's no visible hint of it), this cancel and the close below will not be called - in tests this is annoying because you often get a cascade of failures as ports are not closed correctly, but more generally it's bad because of the proliferation of poor code examples that later are copy-pasted

raising exceptions in nim is very easy and convenient - looking at the outcomes in our code, handling them correctly and consistently turns out to be very hard in practice

Comment (sinkingsugar) [reply tests/testmplex.nim]: Well you are right but I see no fault in the way the tests are written. 
This is more because there is no RAII/Destructors/Defer taking care of the situation behind the scene.

Altho isn't this supposed to take care of the scenario you gave? If not I guess we should try/catch/finally every test.
https://github.com/status-im/nim-libp2p/blob/b3f74ba91c23510242be178bf57d08f6665eb5b7/tests/testmplex.nim#L176-L177

Anyway in tests you want to have early failure with constructs like `check`

Comment (arnetheduck) [reply tests/testmplex.nim]: > This is more because there is no RAII/Destructors/Defer taking care of the situation behind the scene.

well, that's the point kind of - if you want to use exceptions in nim correctly, you have to write `try`/`finally`/`defer` - also in tests - else the tests leak resources and are no longer independent - unfortunately, this kind of code is not unusual at all (in fact, both our and standard library code is riddled with leaks like these), tests or "production" code alike. 

In this particular case, the right thing would be be `defer: close` - as for the cancellation, I don't remember if future supports cancelling already-triggered futures - if not, it's defer+if and a flag..

Comment (sinkingsugar) [reply tests/testmplex.nim]: Yeah I was thinking to use `defer` but wasn't `defer` broken 😄 ? I refrained using it kinda.. maybe it's fine now.
Anyway I will have a look at it!

Comment (arnetheduck) [reply tests/testmplex.nim]: it used to be - specially in `async`, and still is, but the situation has improved and some simple cases are handled correctly now :)

**Comment by dryajov** [comment]
> Fixed the issue, also I noticed that all the tests were inconsistent in terms of actually waiting for the async events to be processed... so I'm proposing to use this pattern in every test:
> [sinkingsugar@fb5fcc6](https://github.com/sinkingsugar/nim-libp2p/commit/fb5fcc6c8b4775e9ed4466cbb2066d7f82a403b6)
> 
> Basically an async lock + timeout to actually make sure tests ran.

I'm not sure if I follow, where are they not waiting for async events? We're not waiting for all the events to be processed but just for some condition to be triggered.

**Comment by sinkingsugar** [comment]
They were basically never waiting reliably due to the nature of the async scheduler, you would barely notice if the messages were small because the operation probably was sequential once the async callback triggered. 
But with a big message if you don't properly wait for the condition in the listener long enough it would fail and since pending would be fed into the next test causing read overflow in the next test.

**Comment by dryajov** [comment]
> They were basically never waiting reliably due to the nature of the async scheduler, you would barely notice if the messages were small because the operation probably was sequential once the async callback triggered.

Can you point me to an example? I'm pretty conscious about properly awaiting things where it matters.

**Comment by dryajov** [comment]
One more thing, we need to shutdown the transport once we've finished with the test.

File: tests/testmplex.nim [hunk 3]
```diff
@@ -140,6 +144,48 @@ suite "Mplex":
       let mplexDial = newMplex(conn)
       let stream  = await mplexDial.newStream()
       await stream.writeLp("Hello from stream!")
+      await lock or timeout
```
Comment (dryajov) [inline tests/testmplex.nim]: I would actually use `wait` from `nim-chronos` here instead, if it timed out then wait throws an exception, `nim-unittest` has an `expect` clause to check for expected exceptions and we can cleanup in the finally clause of our `try`.

Comment (sinkingsugar) [reply tests/testmplex.nim]: > > They were basically never waiting reliably due to the nature of the async scheduler, you would barely notice if the messages were small because the operation probably was sequential once the async callback triggered.
> 
> Can you point me to an example? I'm pretty conscious about properly awaiting things where it matters.

For example doing like this:
https://github.com/sinkingsugar/nim-libp2p/commit/d6813775352e4bbdf5b5cea6de50d72a85200197
Btw, there is nothing really wrong, it's just error prone with delays (and sending lot of data gets split into different async scheduler iterations likely?)
Basically anything that happens inside `handleMplexListen` is not guaranteed to happen, and so that `lock` (probably bad name) waits or timeouts for that.

Btw nice catch on closing transports, that is likely why my big buffer was getting pushed into the next test 😸

Comment (sinkingsugar) [reply tests/testmplex.nim]: Btw I'm thinking to do the transport closing as part of https://github.com/status-im/nim-libp2p/issues/75

**Comment by sinkingsugar** [comment]
@dryajov btw since there is already a receive limit of 1MB here:
https://github.com/sinkingsugar/nim-libp2p/blob/1e393fb3b56c418d3220c83340492d7304620f0b/libp2p/connection.nim#L127
Shouldn't we just put a send limit of 1MB in the proc below and that's it rather than mplex?
Altho if we go by spec things are tricky.. I see no limits at all except mplex (so the receive limit is not exactly correct maybe?)

**Comment by sinkingsugar** [comment]
Small heads up:
I found a new issue where some big payload won't go thru and reach the actual mplex channel stream
image
Might be a problem within chronos streams by the look of it.

**Comment by sinkingsugar** [comment]
Since it seems that the above is a issue in chronos:
https://github.com/status-im/nim-chronos/issues/72
I'm proposing we actually put a hard limit in `connection.nim` indeed:
https://github.com/sinkingsugar/nim-libp2p/blob/1e393fb3b56c418d3220c83340492d7304620f0b/libp2p/connection.nim#L141

File: tests/testmplex.nim (hunk 3 above)
Comment (dryajov) [reply tests/testmplex.nim]: SGTM, lets move cleanup of things to a central location.

**Comment by sinkingsugar** [comment]
Going to re-open from main repo since I cannot move it.
